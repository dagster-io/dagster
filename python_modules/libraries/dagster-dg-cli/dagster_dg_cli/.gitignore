docs/packages
docs/node_modules
