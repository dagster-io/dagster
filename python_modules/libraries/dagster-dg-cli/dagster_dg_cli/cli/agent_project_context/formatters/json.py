"""JSON formatter for project context."""

import json

from dagster_dg_cli.cli.agent_project_context.formatters import BaseFormatter
from dagster_dg_cli.cli.agent_project_context.models import ProjectContext


class JsonFormatter(BaseFormatter):
    """Formats project context as JSON."""

    def format(self, context: ProjectContext) -> str:
        """Format project context as JSON."""
        # Convert record objects to dictionaries for JSON serialization
        context_dict = {
            "working_directory": context.working_directory,
            "project_root": context.project_root,
            "project_name": context.project_name,
            "components": [
                {
                    "key": comp.key,
                    "summary": comp.summary,
                    "namespace": comp.namespace,
                    "name": comp.name,
                }
                for comp in context.components
            ],
            "integrations": [
                {
                    "name": integ.name,
                    "description": integ.description,
                    "pypi": integ.pypi,
                }
                for integ in context.integrations
            ],
            "project_structure": {
                "has_pyproject_toml": context.project_structure.has_pyproject_toml,
                "has_dbt_project": context.project_structure.has_dbt_project,
                "dbt_project_paths": context.project_structure.dbt_project_paths,
                "python_packages": context.project_structure.python_packages,
            },
            "dbt_schemas": {
                key: {
                    "description": schema.description,
                    "scaffold_params_schema": schema.scaffold_params_schema,
                    "component_schema": schema.component_schema,
                }
                for key, schema in context.dbt_schemas.items()
            },
        }
        return json.dumps(context_dict, indent=2)
