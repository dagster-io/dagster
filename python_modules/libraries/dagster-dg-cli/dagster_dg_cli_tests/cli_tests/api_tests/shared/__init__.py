"""Shared utilities for API test recordings."""
