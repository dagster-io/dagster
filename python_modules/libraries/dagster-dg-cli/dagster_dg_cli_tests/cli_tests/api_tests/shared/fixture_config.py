"""Configuration for API fixture recording scenarios."""

from dataclasses import dataclass
from typing import Optional


@dataclass
class FixtureScenario:
    """Configuration for a single fixture recording scenario."""

    command: str
    depends_on: Optional[list[str]] = None  # Dependencies (run IDs, etc.)
