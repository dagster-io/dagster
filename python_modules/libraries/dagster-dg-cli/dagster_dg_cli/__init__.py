from dagster_shared.libraries import DagsterLibraryRegistry

from dagster_dg_cli.version import __version__ as __version__

DagsterLibraryRegistry.register("dagster-dg-cli", __version__)
