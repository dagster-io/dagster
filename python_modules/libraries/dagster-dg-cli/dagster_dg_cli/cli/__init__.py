import os
from pathlib import Path

import click
from dagster_dg_core.config import normalize_cli_config
from dagster_dg_core.context import DgContext
from dagster_dg_core.shared_options import dg_global_options, dg_path_options
from dagster_dg_core.utils import DG_CLI_MAX_OUTPUT_WIDTH, DgClickGroup

from dagster_dg_cli.cli.check import check_group
from dagster_dg_cli.cli.dev import dev_command
from dagster_dg_cli.cli.docs import docs_group
from dagster_dg_cli.cli.launch import launch_command
from dagster_dg_cli.cli.list import list_group
from dagster_dg_cli.cli.mcp_server import mcp_group
from dagster_dg_cli.cli.plus import plus_group
from dagster_dg_cli.cli.scaffold import scaffold_group
from dagster_dg_cli.cli.utils import utils_group
from dagster_dg_cli.version import __version__


def create_dg_cli():
    @click.group(
        name="dg",
        commands={
            "check": check_group,
            "docs": docs_group,
            "utils": utils_group,
            "launch": launch_command,
            "list": list_group,
            "scaffold": scaffold_group,
            "dev": dev_command,
            "plus": plus_group,
            "mcp": mcp_group,
        },
        context_settings={
            "max_content_width": DG_CLI_MAX_OUTPUT_WIDTH,
            "help_option_names": ["-h", "--help"],
        },
        invoke_without_command=True,
        cls=DgClickGroup,
    )
    @dg_path_options
    @dg_global_options
    @click.option(
        "--clear-cache",
        is_flag=True,
        help="Clear the cache.",
        default=False,
    )
    @click.option(
        "--install-completion",
        is_flag=True,
        help="Automatically detect your shell and install a completion script for the `dg` command. This will append to your shell startup file.",
        default=False,
    )
    @click.version_option(__version__, "--version", "-v")
    def group(
        install_completion: bool,
        clear_cache: bool,
        path: Path,
        **global_options: object,
    ):
        """CLI for managing Dagster projects."""
        os.environ["DAGSTER_IS_DEV_CLI"] = "1"

        context = click.get_current_context()
        if install_completion:
            import dagster_dg_core.completion

            dagster_dg_core.completion.install_completion(context)
            context.exit(0)
        elif clear_cache:
            from dagster_dg_core.cache import DgCache

            cli_config = normalize_cli_config(global_options, context)
            dg_context = DgContext.from_file_discovery_and_command_line_config(path, cli_config)
            # Normally we would access the cache through the DgContext, but cache is currently
            # disabled outside of a project context. When that restriction is lifted, we will change
            # this to access the cache through the DgContext.
            cache = DgCache.from_config(dg_context.config)
            cache.clear_all()
            if context.invoked_subcommand is None:
                context.exit(0)

        elif context.invoked_subcommand is None:
            click.echo(context.get_help())
            context.exit(0)

    return group


ENV_PREFIX = "DAGSTER_DG"
cli = create_dg_cli()


def main():
    cli(auto_envvar_prefix=ENV_PREFIX)
