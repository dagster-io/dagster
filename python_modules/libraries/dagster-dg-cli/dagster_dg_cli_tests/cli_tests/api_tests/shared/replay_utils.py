"""Shared utilities for replaying recorded GraphQL responses in tests."""

import json
from collections.abc import Mapping
from pathlib import Path
from typing import Any, Optional

from dagster_dg_cli.utils.plus.gql_client import DagsterPlusGraphQLClient


class ReplayClient(DagsterPlusGraphQLClient):
    """GraphQL client that replays recorded responses."""

    def __init__(self, responses: list[dict[str, Any]]):
        # Don't call super().__init__ as we don't need the real GraphQL client
        self.responses = responses
        self.call_index = 0

    def execute(self, query: str, variables: Optional[Mapping[str, Any]] = None) -> dict:
        """Return next recorded response."""
        if self.call_index >= len(self.responses):
            raise ValueError(f"Exhausted {len(self.responses)} responses")
        response = self.responses[self.call_index]
        self.call_index += 1
        return response


def load_recorded_graphql_responses(
    recordings_dir: Path, scenario_name: str
) -> list[dict[str, Any]]:
    """Load GraphQL response recordings for a given scenario.

    Args:
        recordings_dir: Directory containing the recordings
        scenario_name: Name of the scenario to load

    Returns:
        List of recorded GraphQL responses
    """
    scenario_folder = recordings_dir / scenario_name

    if not scenario_folder.exists():
        raise ValueError(f"Recording scenario not found: {scenario_folder}")

    json_files = sorted([f for f in scenario_folder.glob("*.json") if f.name[0:2].isdigit()])

    if not json_files:
        raise ValueError(f"No numbered JSON files found in {scenario_folder}")

    responses = []
    for json_file in json_files:
        with open(json_file) as f:
            responses.append(json.load(f))

    return responses
