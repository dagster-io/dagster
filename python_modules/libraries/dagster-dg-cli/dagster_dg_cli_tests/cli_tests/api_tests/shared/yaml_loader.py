"""YAML loading utilities for fixture scenario configurations."""

from pathlib import Path

import yaml

from dagster_dg_cli_tests.cli_tests.api_tests.shared.fixture_config import FixtureScenario


def load_fixture_scenarios_from_yaml(yaml_file: Path) -> dict[str, FixtureScenario]:
    """Load fixture scenarios from a YAML file.

    Args:
        yaml_file: Path to YAML file containing fixture scenario definitions

    Returns:
        Dictionary mapping fixture names to FixtureScenario objects

    Raises:
        ValueError: If YAML file doesn't exist or has invalid format
    """
    if not yaml_file.exists():
        raise ValueError(f"YAML fixture file not found: {yaml_file}")

    try:
        with open(yaml_file) as f:
            yaml_data = yaml.safe_load(f) or {}
    except yaml.YAMLError as e:
        raise ValueError(f"Invalid YAML in {yaml_file}: {e}")

    fixture_scenarios = {}
    for fixture_name, config_data in yaml_data.items():
        if not isinstance(config_data, dict):
            raise ValueError(
                f"Invalid fixture config for '{fixture_name}': expected dict, got {type(config_data)}"
            )

        if "command" not in config_data:
            raise ValueError(f"Missing 'command' field for fixture '{fixture_name}'")

        fixture_scenarios[fixture_name] = FixtureScenario(
            command=config_data["command"], depends_on=config_data.get("depends_on")
        )

    return fixture_scenarios
