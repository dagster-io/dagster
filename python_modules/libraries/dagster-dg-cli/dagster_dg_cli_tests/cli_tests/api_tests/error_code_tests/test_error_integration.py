"""End-to-end error integration testing.

This module tests complete error scenarios from GraphQL responses through the API layer
to CLI output, ensuring proper error propagation and handling across all API domains.
"""

import json
from pathlib import Path
from typing import Any

import pytest
import yaml
from click.testing import CliRunner
from dagster_dg_cli.cli import cli as root_cli
from dagster_dg_cli.cli.api.client import DgApiTestContext
from syrupy.assertion import SnapshotAssertion

from dagster_dg_cli_tests.cli_tests.api_tests.shared.replay_utils import ReplayClient


class TestErrorIntegration:
    """Test end-to-end error handling across all API domains."""

    @pytest.fixture
    def scenarios(self) -> dict[str, Any]:
        """Load error code test scenarios."""
        current_dir = Path(__file__).parent
        scenarios_file = current_dir / "scenarios.yaml"

        with open(scenarios_file) as f:
            scenarios = yaml.safe_load(f) or {}

        return scenarios

    @pytest.fixture
    def asset_error_scenarios(self, scenarios: dict[str, Any]) -> dict[str, Any]:
        """Filter scenarios that test asset API errors."""
        return {
            name: scenario
            for name, scenario in scenarios.items()
            if "dg api asset" in scenario["command"]
        }

    @pytest.fixture
    def deployment_error_scenarios(self, scenarios: dict[str, Any]) -> dict[str, Any]:
        """Filter scenarios that test deployment API errors."""
        return {
            name: scenario
            for name, scenario in scenarios.items()
            if "dg api deployment" in scenario["command"]
        }

    def test_complete_error_flow_json_output(
        self, scenarios: dict[str, Any], snapshot: SnapshotAssertion
    ):
        """Test complete error flow from GraphQL to JSON CLI output."""
        json_scenarios = {
            name: scenario
            for name, scenario in scenarios.items()
            if "--json" in scenario["command"]
        }

        for scenario_name, scenario in json_scenarios.items():
            # Load recorded responses for this scenario
            recordings_dir = Path(__file__).parent / "recordings"
            scenario_folder = recordings_dir / scenario_name

            # Skip if recording doesn't exist
            if not scenario_folder.exists():
                pytest.skip(
                    f"Recording not found for {scenario_name}. Run: make record SCENARIO={scenario_name}"
                )

            # Load all recorded responses
            from dagster_dg_cli_tests.cli_tests.api_tests.shared.replay_utils import (
                load_recorded_graphql_responses,
            )

            try:
                recorded_responses = load_recorded_graphql_responses(recordings_dir, scenario_name)
            except ValueError:
                pytest.skip(f"No valid recordings found for {scenario_name}")

            # Create test context with recorded responses
            replay_client = ReplayClient(recorded_responses)
            test_context = DgApiTestContext(client_factory=lambda config: replay_client)

            # Execute command
            runner = CliRunner()
            command_args = scenario["command"].split()[1:]  # Remove 'dg'
            result = runner.invoke(root_cli, command_args, obj=test_context, catch_exceptions=False)

            # Create snapshot data
            snapshot_data = {
                "scenario": scenario_name,
                "command": scenario["command"],
                "exit_code": result.exit_code,
                "output": result.output,
                "expected_code": scenario.get("expected_code"),
                "expected_status": scenario.get("expected_status"),
                "description": scenario.get("description"),
            }

            # Assert against snapshot
            assert snapshot_data == snapshot(name=f"json_error_flow_{scenario_name}")

    def test_complete_error_flow_text_output(
        self, scenarios: dict[str, Any], snapshot: SnapshotAssertion
    ):
        """Test complete error flow from GraphQL to text CLI output."""
        text_scenarios = {
            name: scenario
            for name, scenario in scenarios.items()
            if "--json" not in scenario["command"]
        }

        for scenario_name, scenario in text_scenarios.items():
            # Load recorded responses for this scenario
            recordings_dir = Path(__file__).parent / "recordings"
            scenario_folder = recordings_dir / scenario_name

            # Skip if recording doesn't exist
            if not scenario_folder.exists():
                pytest.skip(
                    f"Recording not found for {scenario_name}. Run: make record SCENARIO={scenario_name}"
                )

            # Load all recorded responses
            from dagster_dg_cli_tests.cli_tests.api_tests.shared.replay_utils import (
                load_recorded_graphql_responses,
            )

            try:
                recorded_responses = load_recorded_graphql_responses(recordings_dir, scenario_name)
            except ValueError:
                pytest.skip(f"No valid recordings found for {scenario_name}")

            # Create test context with recorded responses
            replay_client = ReplayClient(recorded_responses)
            test_context = DgApiTestContext(client_factory=lambda config: replay_client)

            # Execute command
            runner = CliRunner()
            command_args = scenario["command"].split()[1:]  # Remove 'dg'
            result = runner.invoke(root_cli, command_args, obj=test_context, catch_exceptions=False)

            # Create snapshot data
            snapshot_data = {
                "scenario": scenario_name,
                "command": scenario["command"],
                "exit_code": result.exit_code,
                "output": result.output,
                "expected_code": scenario.get("expected_code"),
                "expected_status": scenario.get("expected_status"),
                "description": scenario.get("description"),
            }

            # Assert against snapshot
            assert snapshot_data == snapshot(name=f"text_error_flow_{scenario_name}")

    def test_cross_domain_error_consistency(self, scenarios: dict[str, Any]):
        """Test that similar errors are handled consistently across different API domains."""
        # Group scenarios by error type
        error_groups = {}

        for scenario_name, scenario in scenarios.items():
            expected_code = scenario.get("expected_code")
            if expected_code:
                if expected_code not in error_groups:
                    error_groups[expected_code] = []
                error_groups[expected_code].append((scenario_name, scenario))

        # Test consistency within each error group
        for error_code, scenarios_in_group in error_groups.items():
            if len(scenarios_in_group) < 2:
                continue  # Need at least 2 scenarios to test consistency

            # Check that all scenarios with same error code have same status
            expected_statuses = set()
            for scenario_name, scenario in scenarios_in_group:
                expected_status = scenario.get("expected_status")
                if expected_status:
                    expected_statuses.add(expected_status)

            assert len(expected_statuses) <= 1, (
                f"Error code {error_code} has inconsistent status codes: {expected_statuses}"
            )

    def test_error_propagation_through_api_layers(self, asset_error_scenarios: dict[str, Any]):
        """Test that errors properly propagate through API layers without mutation."""
        for scenario_name, scenario in asset_error_scenarios.items():
            # Load recorded responses
            recordings_dir = Path(__file__).parent / "recordings"

            if not (recordings_dir / scenario_name).exists():
                continue

            try:
                from dagster_dg_cli_tests.cli_tests.api_tests.shared.replay_utils import (
                    load_recorded_graphql_responses,
                )

                recorded_responses = load_recorded_graphql_responses(recordings_dir, scenario_name)
            except (ValueError, IndexError):
                continue

            # Test that GraphQL errors in response are properly handled
            if "errors" in recorded_responses[0]:
                # Create test context
                replay_client = ReplayClient(recorded_responses)
                test_context = DgApiTestContext(client_factory=lambda config: replay_client)

                # Execute command
                runner = CliRunner()
                command_args = scenario["command"].split()[1:]  # Remove 'dg'
                result = runner.invoke(
                    root_cli, command_args, obj=test_context, catch_exceptions=False
                )

                # Should exit with error
                assert result.exit_code != 0, (
                    f"Command should fail when GraphQL errors present in {scenario_name}"
                )

                # If JSON output, should have proper error structure
                if "--json" in scenario["command"]:
                    try:
                        error_response = json.loads(result.output)
                        assert "error" in error_response, (
                            f"JSON error response should have 'error' field in {scenario_name}"
                        )
                        assert "code" in error_response, (
                            f"JSON error response should have 'code' field in {scenario_name}"
                        )
                    except json.JSONDecodeError:
                        pytest.fail(f"Invalid JSON output for {scenario_name}")

    def test_fallback_error_handling(self, scenarios: dict[str, Any]):
        """Test fallback behavior for unknown or malformed error responses."""
        # Test with completely empty response
        empty_response = {}

        # Pick a representative scenario
        representative_scenario = None
        for scenario_name, scenario in scenarios.items():
            if "--json" in scenario["command"]:
                representative_scenario = (scenario_name, scenario)
                break

        if not representative_scenario:
            pytest.skip("No JSON scenario found for fallback testing")

        scenario_name, scenario = representative_scenario

        # Create test context with empty response
        replay_client = ReplayClient([empty_response])
        test_context = DgApiTestContext(client_factory=lambda config: replay_client)

        # Execute command
        runner = CliRunner()
        command_args = scenario["command"].split()[1:]  # Remove 'dg'

        # This should not crash, should handle gracefully
        try:
            result = runner.invoke(root_cli, command_args, obj=test_context, catch_exceptions=False)
            # Should exit with some kind of error code
            assert result.exit_code != 0, "Empty response should result in error"
        except Exception:
            # If it does crash, this is a test failure - fallback should handle gracefully
            pytest.fail("Command crashed on empty response - fallback handling insufficient")

    def test_error_message_localization_consistency(self, scenarios: dict[str, Any]):
        """Test that error messages are consistent and properly localized."""
        for scenario_name, scenario in scenarios.items():
            # Load recorded responses
            recordings_dir = Path(__file__).parent / "recordings"

            if not (recordings_dir / scenario_name).exists():
                continue

            try:
                from dagster_dg_cli_tests.cli_tests.api_tests.shared.replay_utils import (
                    load_recorded_graphql_responses,
                )

                recorded_responses = load_recorded_graphql_responses(recordings_dir, scenario_name)
            except (ValueError, IndexError):
                continue

            # Create test context
            replay_client = ReplayClient(recorded_responses)
            test_context = DgApiTestContext(client_factory=lambda config: replay_client)

            # Execute command
            runner = CliRunner()
            command_args = scenario["command"].split()[1:]  # Remove 'dg'
            result = runner.invoke(root_cli, command_args, obj=test_context, catch_exceptions=False)

            if result.exit_code == 0:
                continue  # Skip successful commands

            # Error messages should be in English and properly formatted
            output = result.output.strip()
            if output:
                # Should not contain raw GraphQL error markup
                assert "<" not in output, (
                    f"Error message contains markup in {scenario_name}: {output}"
                )
                assert ">" not in output, (
                    f"Error message contains markup in {scenario_name}: {output}"
                )

                # Should not contain internal Python paths or stack traces
                assert "Traceback" not in output, (
                    f"Error message contains traceback in {scenario_name}"
                )
                assert "/python" not in output.lower(), (
                    f"Error message contains path in {scenario_name}"
                )

    def test_concurrent_error_handling(self, scenarios: dict[str, Any]):
        """Test that error handling is consistent under concurrent access patterns."""
        # This is a placeholder for future concurrent testing
        # For now, just ensure error handling doesn't have shared state issues

        json_scenarios = [
            (name, scenario)
            for name, scenario in scenarios.items()
            if "--json" in scenario["command"]
        ][:3]  # Test with first 3 JSON scenarios

        results = []

        for scenario_name, scenario in json_scenarios:
            # Load recorded responses
            recordings_dir = Path(__file__).parent / "recordings"

            if not (recordings_dir / scenario_name).exists():
                continue

            try:
                from dagster_dg_cli_tests.cli_tests.api_tests.shared.replay_utils import (
                    load_recorded_graphql_responses,
                )

                recorded_responses = load_recorded_graphql_responses(recordings_dir, scenario_name)
            except (ValueError, IndexError):
                continue

            # Create test context
            replay_client = ReplayClient(recorded_responses)
            test_context = DgApiTestContext(client_factory=lambda config: replay_client)

            # Execute command
            runner = CliRunner()
            command_args = scenario["command"].split()[1:]  # Remove 'dg'
            result = runner.invoke(root_cli, command_args, obj=test_context, catch_exceptions=False)

            results.append((scenario_name, result.exit_code, result.output))

        # Each execution should be independent - no shared state corruption
        for i, (scenario_name, exit_code, output) in enumerate(results):
            assert exit_code != 0, f"Error scenario {scenario_name} should fail"
            assert output.strip(), f"Error scenario {scenario_name} should have output"
