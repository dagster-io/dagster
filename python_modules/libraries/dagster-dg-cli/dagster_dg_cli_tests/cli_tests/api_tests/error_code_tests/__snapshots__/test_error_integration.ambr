# serializer version: 1
# name: TestErrorIntegration.test_complete_error_flow_json_output[json_error_flow_error_invalid_asset_key_empty]
  dict({
    'command': "dg api asset get '' --json",
    'description': 'Empty asset key should trigger validation error',
    'exit_code': 5,
    'expected_code': 'INVALID_SUBSET',
    'expected_status': 400,
    'output': '''
      {"error": "Asset not found: ''", "code": "INTERNAL_ERROR", "statusCode": 500, "type": "server_error"}
  
    ''',
    'scenario': 'error_invalid_asset_key_empty',
  })
# ---
# name: TestErrorIntegration.test_complete_error_flow_json_output[json_error_flow_error_invalid_asset_key_malformed]
  dict({
    'command': "dg api asset get 'invalid::asset::key' --json",
    'description': 'Malformed asset key with invalid characters',
    'exit_code': 5,
    'expected_code': 'INVALID_SUBSET',
    'expected_status': 400,
    'output': '''
      {"error": "Asset not found: 'invalid::asset::key'", "code": "INTERNAL_ERROR", "statusCode": 500, "type": "server_error"}
  
    ''',
    'scenario': 'error_invalid_asset_key_malformed',
  })
# ---
# name: TestErrorIntegration.test_complete_error_flow_text_output[text_error_flow_error_asset_not_found_text]
  dict({
    'command': 'dg api asset get nonexistent-asset-key',
    'description': 'Asset not found error in text format',
    'exit_code': 1,
    'expected_code': 'ASSET_NOT_FOUND',
    'expected_status': 404,
    'output': '''
      Error querying Dagster Plus API: Asset not found: nonexistent-asset-key
  
    ''',
    'scenario': 'error_asset_not_found_text',
  })
# ---
# name: TestErrorIntegration.test_complete_error_flow_text_output[text_error_flow_error_unauthorized_text]
  dict({
    'command': 'dg api asset list',
    'description': 'Authentication error in text format',
    'exit_code': 0,
    'expected_code': 'UNAUTHORIZED',
    'expected_status': 401,
    'output': '''
      Asset Key: aws/cloud-prod/user_roles
      ID: dagster_open_platform.__repository__.["aws", "cloud-prod", "user_roles"]
      Description: Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.
      Group: aws_stages
      Kinds: None
      
      Asset Key: aws/cloud-prod/workspace_staging_asset_checks
      ID: dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_asset_checks"]
      Description: Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.
      Group: aws_stages
      Kinds: None
      
      Asset Key: aws/cloud-prod/workspace_staging_assets
      ID: dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_assets"]
      Description: Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.
      Group: aws_stages
      Kinds: None
      
      Asset Key: aws/cloud-prod/workspace_staging_external_repo_metadata
      ID: dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_external_repo_metadata"]
      Description: Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.
      Group: aws_stages
      Kinds: None
      
      Asset Key: aws/cloud-prod/workspace_staging_jobs
      ID: dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_jobs"]
      Description: Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.
      Group: aws_stages
      Kinds: None
      
      Asset Key: aws/cloud-prod/workspace_staging_metadata
      ID: dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_metadata"]
      Description: Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.
      Group: aws_stages
      Kinds: None
      
      Asset Key: aws/cloud-prod/workspace_staging_partitions
      ID: dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_partitions"]
      Description: Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.
      Group: aws_stages
      Kinds: None
      
      Asset Key: aws/cloud-prod/workspace_staging_pipelines
      ID: dagster_open_platform.__repository__.["aws", "cloud-prod", "workspace_staging_pipelines"]
      Description: Snowflake stages for AWS data, creates new stages for new assets, refreses existing stages.
      Group: aws_stages
      Kinds: None
  
    ''',
    'scenario': 'error_unauthorized_text',
  })
# ---
